spec/Candid.md

```diff
@@ -864,143 +872,200 @@ service { <methtype>;* } <: service { <methtype'>;* }
 service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>; <methtype'>;* }
 ```
 
-### Elaboration
+### Coercion
+
+This subtyping is implemented during the deserialization of Candid at an expected type. As described in [Section Deserialisation](#deserialization), the binary value is conceptually first _decoded_ into an abstract value, and then _coerced_ to the expected type.
+
+This section describes the covercion, as a ternary relation `V ~> V' : T` to describe when a value `V` can be coerced to a value `V'` of type `T`. The fields `V` and `T` can be understood as inputs and `V'` as the output of this relation.
 
-To define the actual coercion function used during deserialisation of a value, we extend the subtyping relation to a ternary *elaboration* relation `T <: T' ~> f`, where `f` is a suitable coercion function of type `T -> T'`.
+Here `V` models untyped values, which form an abstract data model of both the message in transit (`V`), as well as the result of the coercion (`V'`). In the following, we re-use the syntax of the textual representation, with the following changes to make it free of overloading:
 
+ * Number literals (`<primval>`) must be immediately enclosed with an `<annval>` that clarifies the precise number type.
+ * The canonical value of type `reserved` is expressed as `(null : reserved)`.
+ * No other use of `<annval>`.
```

Comment:
We also need annval for `null : opt _`

Reply:
No, I am pretty sure we don't! That’s because 
 * On the left of `~>`, it doesn't matter which `null` it is (any `null` behaves the same) 
 * On the right of `~>`, the type is given, so again there is only no `null` value.

So it is fine to have a single `null` value in the `V` set.

This is not true for overloaded numbers; `(5 : nat) ~>`  and `(5 : int) ~>` behave differently.

Reply:
In other words, `null` has a principal type, numbers don't. (Which makes me wonder, why do we need a special case for `reserved`?)

Reply:
Because the algebra of _binary_ values has a dedicated `reserved` value (`DIDL\x00\x01\x70`), but our textual representation doesn’t. The `(null : reserved)` could just as well be `("reserved" : reserved)` or `(record {} : reserved)`; it’s arbitrary and, in a way, always lie.

For me, the conceptual “abstract value data model” _has_ a dedicated `reserved` value, and I think it would be good to have it in the textual representation as well, to avoid the above wart in the formalism, but also to not force debug output to come up with such arbitrary values when decoding a message with a `reserved` value:
```
~/dfinity/haskell-candid $ cabal -v0 run hcandid -- --decode --rust "DIDL\x00\x01\x70"
((null : reserved))
```

Reply:
More evidence that it might be useful to have it: All implementations that have a “value” type seemed to feel the need to add it to their data type:
* https://github.com/dfinity/candid/blob/03773cc2c3d86b898d78e3161be3c6bae34dbd97/rust/candid/src/parser/value.rs#L37
* https://github.com/nomeata/haskell-candid/blob/36202a9aef4387925c0d0de23a0e3645ed7d2b3e/src/Codec/Candid/Types.hs#L142